sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala

```diff
@@ -2721,6 +2721,16 @@ object SQLConf {
       .booleanConf
       .createWithDefault(false)
 
+  val UNION_BYNAME_STRUCT_SUPPORT_ENABLED =
+    buildConf("spark.sql.unionByName.structSupport.enabled")
+      .doc("When true, the `allowMissingColumns` feature of `Dataset.unionByName` supports " +
+        "nested column in struct types. Missing nested columns of struct columns with same " +
+        "name will also be filled with null values. This currently does not support nested " +
+        "columns in array and map types.")
```

Comment:
How about explaining the behavior of the fields being sorted when merging them? I'm a bit worried that users might be surprised about the behavior.
```
scala> val df1 = spark.range(1).selectExpr("id c0", "named_struct('c', id + 1, 'b', id + 2, 'a', id + 3) c1")
scala> val df2 = spark.range(1).selectExpr("id c0", "named_struct('c', id + 1, 'b', id + 2) c1")

scala> df1.unionByName(df1, true).printSchema()
root
 |-- c0: long (nullable = false)
 |-- c1: struct (nullable = false)
 |    |-- c: long (nullable = false)
 |    |-- b: long (nullable = false)
 |    |-- a: long (nullable = false)

scala> df1.unionByName(df2, true).printSchema()
root
 |-- c0: long (nullable = false)
 |-- c1: struct (nullable = false)
 |    |-- a: long (nullable = true)
 |    |-- b: long (nullable = false)
 |    |-- c: long (nullable = false)
```

Reply:
Yeah .. I was thinking about same thing.

Reply:
Oh, right, I will update the doc.

Reply:
Updated both the config doc and API doc in Dataset.

Reply:
Thanks for the update. Btw, to follow a top-level case (See an example below), is it technically difficult to reorder struct fields by adding `Project` on the top of `Union`? Does this make the logic ugly?
```
// The top-level case
scala> val df1 = spark.range(1).selectExpr("id c", "id a")
scala> val df2 = spark.range(1).selectExpr("id c", "id b")
scala> df1.unionByName(df2, true).printSchema
root
 |-- c: long (nullable = false)
 |-- a: long (nullable = true)
 |-- b: long (nullable = true)

// The struct case (the current behaviour)
scala> val df3 = spark.range(1).selectExpr("named_struct('c', id, 'a', id) col")
scala> val df4 = spark.range(1).selectExpr("named_struct('c', id, 'b', id) col")
scala> df3.unionByName(df4, true).printSchema
root
 |-- col: struct (nullable = false)
 |    |-- a: long (nullable = true)
 |    |-- b: long (nullable = true)
 |    |-- c: long (nullable = false)

// The following column order is more natural?
scala> df3.unionByName(df4, true).printSchema
root
 |-- col: struct (nullable = false)
 |    |-- c: long (nullable = true)
 |    |-- a: long (nullable = true)
 |    |-- b: long (nullable = false)

== Analyzed Logical Plan ==
col: struct<c:bigint,a:bigint,b:bigint>
Project [struct(c, col#114.c, a, col#114.a, b, col#114.b) AS col#116] <=== Adding `Project` in the end of `ResolveUnion`?
+- Union false, false
   :- Project [named_struct(a, col#87.a, b, null, c, col#87.c) AS col#114]
   :  +- Project [named_struct(c, id#85L, a, id#85L) AS col#87]
   :     +- Range (0, 1, step=1, splits=Some(4))
   +- Project [named_struct(a, null, b, col#91.b, c, col#91.c) AS col#113]
      +- Project [named_struct(c, id#89L, b, id#89L) AS col#91]
         +- Range (0, 1, step=1, splits=Some(4))
```

Reply:
To have a consistent schema but not just sorting the fields, top-level case is much easier. We just iterate each top-level column of left side, and in the end putting the fields only existing in right side at the end of schema. Then it is done.

For nested column case, filling missing nested columns is done by using `WithFields` which basically adds new columns in the end of original schema. To achieve same consistent schema like top-level column case, we either adjust all nested columns later or simply sort nested columns.

I believe it is not unable to adjust the nested columns to have a more natural schema. But I can see this PR grows too big and I would like to cut it early. If you are okay, I can address this issue in other PR after this.

Reply:
> I believe it is not unable to adjust the nested columns to have a more natural schema. But I can see this PR grows too big and I would like to cut it early. If you are okay, I can address this issue in other PR after this. WDYT?

Yea, it looks okay to me. cc: @cloud-fan @HyukjinKwon @dongjoon-hyun